File: odnoklassniki-android-sdk-example/src/main/java/ru/ok/android/sdk/example/MainActivity.java
package ru.ok.android.sdk.example;

import java.io.IOException;
import java.util.Currency;
import java.util.EnumSet;
import java.util.HashMap;

import org.json.JSONException;
import org.json.JSONObject;

import android.app.Activity;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.Toast;
import ru.ok.android.sdk.OKRestHelper;
import ru.ok.android.sdk.Odnoklassniki;
import ru.ok.android.sdk.OkListener;
import ru.ok.android.sdk.OkRequestMode;
import ru.ok.android.sdk.Shared;
import ru.ok.android.sdk.util.OkAuthType;
import ru.ok.android.sdk.util.OkDevice;
import ru.ok.android.sdk.util.OkScope;

public class MainActivity extends Activity {
    protected static final String APP_ID = "125497344";
    protected static final String APP_KEY = "CBABPLHIABABABABA";
    protected static final String REDIRECT_URL = "okauth://ok125497344";

    protected Odnoklassniki odnoklassniki;

    private View loginView;
    private View formView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        loginView = findViewById(R.id.login_block);

        Button mLoginBtn = (Button) findViewById(R.id.sdk_login_any);
        Button mLoginBtnSso = (Button) findViewById(R.id.sdk_login_sso);
        Button mLoginBtnOAuth = (Button) findViewById(R.id.sdk_login_oauth);
        mLoginBtn.setOnClickListener(new LoginClickListener(OkAuthType.ANY));
        mLoginBtnSso.setOnClickListener(new LoginClickListener(OkAuthType.NATIVE_SSO));
        mLoginBtnOAuth.setOnClickListener(new LoginClickListener(OkAuthType.WEBVIEW_OAUTH));

        formView = findViewById(R.id.sdk_form);
        findViewById(R.id.sdk_get_currentuser).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(final View view) {
                new GetCurrentUserTask().execute();
            }
        });
        findViewById(R.id.sdk_get_friends).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(final View view) {
                new GetFriendsTask().execute();
            }
        });
        findViewById(R.id.sdk_logout).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(final View view) {
                odnoklassniki.clearTokens();
                hideForm();
            }
        });

        findViewById(R.id.sdk_post).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                odnoklassniki.performPosting(MainActivity.this,
                        "{\"media\":[{\"type\":\"text\",\"text\":\"hello world!\"}]}",
                        false, null);
            }
        });
        findViewById(R.id.sdk_app_invite).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                odnoklassniki.performAppInvite(MainActivity.this, null);
            }
        });
        findViewById(R.id.sdk_app_suggest).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                odnoklassniki.performAppSuggest(MainActivity.this, null);
            }
        });
        findViewById(R.id.sdk_send_note).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                new SdkSendNote().execute();
            }
        });
        findViewById(R.id.sdk_report_payment).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                odnoklassniki.reportPayment(Math.random() + "", "6.28", Currency.getInstance("EUR"));
            }
        });

        odnoklassniki = Odnoklassniki.createInstance(this, APP_ID, APP_KEY);
        odnoklassniki.checkValidTokens(new OkListener() {
            @Override
            public void onSuccess(JSONObject json) {
                showForm();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MainActivity.this, String.format("%s: %s", getString(R.string.error), error), Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (Odnoklassniki.getInstance().isActivityRequestOAuth(requestCode)) {
            Odnoklassniki.getInstance().onAuthActivityResult(requestCode, resultCode, data, getAuthListener());
        } else if (Odnoklassniki.getInstance().isActivityRequestViral(requestCode)) {
            Odnoklassniki.getInstance().onActivityResultResult(requestCode, resultCode, data, getToastListener());
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    /**
     * Creates a listener that displays result as a toast message
     */
    @NonNull
    private OkListener getToastListener() {
        return new OkListener() {
            @Override
            public void onSuccess(final JSONObject json) {
                Toast.makeText(MainActivity.this, json.toString(), Toast.LENGTH_LONG).show();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MainActivity.this, String.format("%s: %s", getString(R.string.error), error), Toast.LENGTH_LONG).show();
            }
        };
    }

    /**
     * Creates a listener that is run on OAUTH authorization completion
     */
    @NonNull
    private OkListener getAuthListener() {
        return new OkListener() {
            @Override
            public void onSuccess(final JSONObject json) {
                try {
                    Toast.makeText(MainActivity.this,
                            String.format("access_token: %s", json.getString("access_token")),
                            Toast.LENGTH_SHORT).show();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                showForm();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MainActivity.this,
                        String.format("%s: %s", getString(R.string.error), error),
                        Toast.LENGTH_SHORT).show();
            }
        };
    }

    protected final void showForm() {
        formView.setVisibility(View.VISIBLE);
        loginView.setVisibility(View.GONE);

        testIfInstallationSourceIsOK();
    }

    protected final void hideForm() {
        formView.setVisibility(View.GONE);
        loginView.setVisibility(View.VISIBLE);
    }

    // Using AsyncTask is arbitrary choice
    // Developers should do a better error handling job ;)

    protected final class GetCurrentUserTask extends AsyncTask<Void, Void, String> {
        @Override
        protected String doInBackground(final Void... params) {
            try {
                return odnoklassniki.request("users.getCurrentUser", null, "get");
            } catch (Exception exc) {
                Log.e("Odnoklassniki", "Failed to get current user info", exc);
            }
            return null;
        }

        @Override
        protected void onPostExecute(final String result) {
            if (result != null) {
                Toast.makeText(MainActivity.this, "Get current user result: " + result, Toast.LENGTH_SHORT).show();
            }
        }
    }

    protected final class GetFriendsTask extends AsyncTask<Void, Void, String> {
        @Override
        protected String doInBackground(final Void... params) {
            try {
                return odnoklassniki.request("friends.get", null, "get");
            } catch (Exception exc) {
                Log.e("Odnoklassniki", "Failed to get friends", exc);
            }
            return null;
        }

        @Override
        protected void onPostExecute(final String result) {
            if (result != null) {
                Toast.makeText(MainActivity.this, "Get user friends result: " + result, Toast.LENGTH_SHORT).show();
            }
        }
    }

    protected class LoginClickListener implements OnClickListener {
        private OkAuthType authType;

        public LoginClickListener(OkAuthType authType) {
            this.authType = authType;
        }

        @Override
        public void onClick(final View view) {
            odnoklassniki.requestAuthorization(MainActivity.this, REDIRECT_URL, authType, OkScope.VALUABLE_ACCESS);
        }
    }

    private void testIfInstallationSourceIsOK() {
        new AsyncTask<Void, Void, Integer>() {
            @Override
            protected Integer doInBackground(Void... params) {
                HashMap<String, String> args = new HashMap<>();
                args.put("adv_id", OkDevice.getAdvertisingId(MainActivity.this));
                int result;
                try {
                    result = Integer.parseInt(odnoklassniki.request("sdk.getInstallSource", args, EnumSet.of(OkRequestMode.UNSIGNED)));
                } catch (IOException | NumberFormatException e) {
                    result = -1;
                }
                return result;
            }

            @Override
            protected void onPostExecute(Integer result) {
                Toast.makeText(MainActivity.this,
                        (result > 0 ?
                                "application installation caused by OK app (" + result + ")" :
                                "application is not caused by OK app (" + result + ")"),
                        Toast.LENGTH_SHORT).show();
            }
        }.execute();
    }

    protected final class SdkSendNote extends AsyncTask<Void, Void, String> {
        @Override
        protected String doInBackground(final Void... params) {
            try {
                OKRestHelper helper = new OKRestHelper(odnoklassniki);
                if (!helper.sdkInit(MainActivity.this)) {
                    Toast.makeText(MainActivity.this, "sdk.sendNote: no sdk token available", Toast.LENGTH_SHORT).show();
                    return null;
                }
                JSONObject note = new JSONObject();
                note.put("uid", "12345678");
                note.put("image", "http://domain.tld/img.png");
                note.put("message", "Hello World from Android SDK!");
                note.put("payload", "SGVsbG8gd29ybGQhISE=");
                helper.sdkSendNote(note, new OkListener() {
                    @Override
                    public void onSuccess(JSONObject json) {
                        String text = json == null ? null : json.toString();
                        Toast.makeText(MainActivity.this, "sdk.sendNote: OK " + text, Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onError(String error) {
                        Toast.makeText(MainActivity.this, "sdk.sendNote: ERROR " + error, Toast.LENGTH_SHORT).show();
                    }
                });
            } catch (Exception e) {
                Log.e(Shared.LOG_TAG, "sdk.sendNote " + e.getMessage(), e);
            }
            return null;
        }
    }
}

File: odnoklassniki-android-sdk-example/src/main/java/ru/ok/android/sdk/example/MainActivity.kt
package ru.ok.android.sdk.example

import java.util.Currency

import org.json.JSONException

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View.*
import android.widget.Toast
import ru.ok.android.sdk.Odnoklassniki
import ru.ok.android.sdk.util.OkAuthType
import ru.ok.android.sdk.util.OkScope
import kotlinx.android.synthetic.main.activity_main.*
import ru.ok.android.sdk.ContextOkListener

// -------------- YOUR APP DATA GOES HERE ------------
private const val APP_ID = "125497344"
private const val APP_KEY = "CBABPLHIABABABABA"
private const val REDIRECT_URL = "okauth://ok125497344"
// -------------- YOUR APP DATA ENDS -----------------

class MainActivity : Activity() {

    private lateinit var ok: Odnoklassniki

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // NOTE: application should use just one of the login methods, ANY is preferable
        sdk_login_any.setOnClickListener {
            ok.requestAuthorization(this, REDIRECT_URL, OkAuthType.ANY, OkScope.VALUABLE_ACCESS)
        }
        sdk_login_sso.setOnClickListener {
            ok.requestAuthorization(this, REDIRECT_URL, OkAuthType.NATIVE_SSO, OkScope.VALUABLE_ACCESS)
        }
        sdk_login_oauth.setOnClickListener {
            ok.requestAuthorization(this, REDIRECT_URL, OkAuthType.WEBVIEW_OAUTH, OkScope.VALUABLE_ACCESS)
        }

        sdk_get_currentuser.setOnClickListener {
            ok.requestAsync("users.getCurrentUser", listener = ContextOkListener(this,
                    onSuccess = { _, json -> toast("Get current user result: $json") },
                    onError = { _, err -> toast("Get current user failed: $err") }
            ))
        }
        sdk_get_friends.setOnClickListener {
            ok.requestAsync("friends.get", listener = ContextOkListener(this,
                    onSuccess = { _, json -> toast("Get user friends result: $json") },
                    onError = { _, err -> toast("Failed to get friends: $err") }
            ))
        }
        sdk_logout.setOnClickListener {
            ok.clearTokens()
            showAppData(false)
        }

        sdk_post.setOnClickListener {
            val json = """
                {"media": [
                    {"type":"text", "text":"Hello world!"},
                    {"type":"link", "url":"https://apiok.ru/"},
                    {"type":"app",
                        "text":"Welcome from sample",
                        "images":[{"url":"https://apiok.ru/res/img/main/app_create.png"}],
                        "actions":[{"text": "Play me!", "mark": "play_me_from_app_block"}]
                    }]
                }
            """.trimIndent()
            ok.performPosting(this, json, false, null)
        }
        sdk_app_invite.setOnClickListener { ok.performAppInvite(this) }
        sdk_app_suggest.setOnClickListener { ok.performAppSuggest(this, null) }
        sdk_report_payment.setOnClickListener {
            ok.reportPayment(Math.random().toString() + "", "6.28", Currency.getInstance("EUR"))
        }

        ok = Odnoklassniki(this, APP_ID, APP_KEY)
        ok.checkValidTokens(ContextOkListener(this,
                onSuccess = { _, _ -> showAppData() },
                onError = { _, err -> toast(getString(R.string.error) + ": $err") }
        ))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        when {
            Odnoklassniki.of(this).isActivityRequestOAuth(requestCode) -> {
                // process OAUTH sign-in response
                Odnoklassniki.of(this).onAuthActivityResult(requestCode, resultCode, data, ContextOkListener(this,
                        onSuccess = { _, json ->
                            try {
                                toast(String.format("access_token: %s", json.getString("access_token")))
                                showAppData()
                            } catch (e: JSONException) {
                                toast("unable to parse login request ${e.message}")
                            }
                        },
                        onError = { _, err -> toast(getString(R.string.error) + ": $err") },
                        onCancel = { _, err -> toast(getString(R.string.auth_cancelled) + ": $err") }
                ))
            }
            Odnoklassniki.of(this).isActivityRequestViral(requestCode) -> {
                // process called viral widgets (suggest / invite / post)
                Odnoklassniki.of(this).onActivityResultResult(requestCode, resultCode, data, ContextOkListener(this,
                        onSuccess = { _, json -> toast(json.toString()) },
                        onError = { _, err -> toast(getString(R.string.error) + ": $err") }
                ))
            }
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    private fun showAppData(loggedIn: Boolean = true) {
        sdk_form.visibility = if (loggedIn) VISIBLE else GONE
        login_block.visibility = if (loggedIn) GONE else VISIBLE
    }

    private fun toast(text: String) = Toast.makeText(this, text, Toast.LENGTH_LONG).show()
}
